modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java

```diff
@@ -125,14 +132,39 @@ public <K, V> IndexQueryResult<K, V> queryLocal(
 
                 IgniteBiTuple<K, V> row = currVal;
 
+                if (isRecordable) {
+                    K k = unwrap(row.getKey(), keepBinary);
```

Comment:
Code duplication, you already have the K and V in `onHasNext`. Just move events to this method. Note, that filtered entries should not raise the event

Reply:
Events moved to `onHasNext`.